File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/TVA.Phasors/DigitalDefinitionBase.vb
'*******************************************************************************************************
'  DigitalDefinitionBase.vb - Digital value definition base class
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  02/18/2005 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Runtime.Serialization
Imports System.ComponentModel

Namespace Phasors

    ''' <summary>This class represents the common implementation of the protocol independent definition of a digital value.</summary>
    <CLSCompliant(False), Serializable()> _
    Public MustInherit Class DigitalDefinitionBase

        Inherits ChannelDefinitionBase
        Implements IDigitalDefinition

        Protected Sub New()
        End Sub

        Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)

            MyBase.New(info, context)

        End Sub

        Protected Sub New(ByVal parent As IConfigurationCell)

            MyBase.New(parent)

        End Sub

        Protected Sub New(ByVal parent As IConfigurationCell, ByVal index As Int32, ByVal label As String)

            MyBase.New(parent, index, label, 1, 0)

        End Sub

        Protected Sub New(ByVal parent As IConfigurationCell, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

            MyBase.New(parent, binaryImage, startIndex)

        End Sub

        ' Derived classes are expected to expose a Public Sub New(ByVal digitalDefinition As IDigitalDefinition)
        Protected Sub New(ByVal digitalDefinition As IDigitalDefinition)

            MyClass.New(digitalDefinition.Parent, digitalDefinition.Index, digitalDefinition.Label)

        End Sub

        <EditorBrowsable(EditorBrowsableState.Never)> _
        Public Overrides ReadOnly Property DataFormat() As DataFormat
            Get
                Return TVA.Phasors.DataFormat.FixedInteger
            End Get
        End Property

        <EditorBrowsable(EditorBrowsableState.Never)> _
        Public NotOverridable Overrides Property Offset() As Single
            Get
                Return MyBase.Offset
            End Get
            Set(ByVal value As Single)
                If value = 0 Then
                    MyBase.Offset = value
                Else
                    Throw New NotImplementedException("Digital values represent bit flags and thus do not support an offset")
                End If
            End Set
        End Property

        <EditorBrowsable(EditorBrowsableState.Never)> _
        Public NotOverridable Overrides Property ScalingFactor() As Int32
            Get
                Return MyBase.ScalingFactor
            End Get
            Set(ByVal value As Int32)
                If value = 1 Then
                    MyBase.ScalingFactor = value
                Else
                    Throw New NotImplementedException("Digital values represent bit flags and thus are not scaled")
                End If
            End Set
        End Property

    End Class

End Namespace

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/DigitalDefinitionBase.cs
//*******************************************************************************************************
//  DigitalDefinitionBase.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  02/18/2005 - James R Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

using System;
using System.ComponentModel;
using System.Runtime.Serialization;

namespace TVA.PhasorProtocols
{
    /// <summary>
    /// Represents the common implementation of the protocol independent definition of a <see cref="IDigitalValue"/>.
    /// </summary>
    [Serializable()]
    public abstract class DigitalDefinitionBase : ChannelDefinitionBase, IDigitalDefinition
    {
        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="DigitalDefinitionBase"/>.
        /// </summary>
        /// <param name="parent">The <see cref="IConfigurationCell"/> parent of this <see cref="DigitalDefinitionBase"/>.</param>
        protected DigitalDefinitionBase(IConfigurationCell parent)
            : base(parent)
        {
            ScalingValue = 1;
            Offset = 0.0D;
        }

        /// <summary>
        /// Creates a new <see cref="DigitalDefinitionBase"/> from specified parameters.
        /// </summary>
        /// <param name="parent">The <see cref="IConfigurationCell"/> parent of this <see cref="DigitalDefinitionBase"/>.</param>
        /// <param name="label">The label of this <see cref="DigitalDefinitionBase"/>.</param>
        protected DigitalDefinitionBase(IConfigurationCell parent, string label)
            : base(parent, label, 1, 0.0D)
        {
        }

        /// <summary>
        /// Creates a new <see cref="DigitalDefinitionBase"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected DigitalDefinitionBase(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets the <see cref="PhasorProtocols.DataFormat"/> of this <see cref="DigitalDefinitionBase"/>.
        /// </summary>
        /// <remarks>
        /// Data format for digital values will always be <see cref="PhasorProtocols.DataFormat.FixedInteger"/>.
        /// </remarks>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public sealed override DataFormat DataFormat
        {
            get
            {
                return PhasorProtocols.DataFormat.FixedInteger;
            }
        }

        /// <summary>
        /// Gets or sets the offset of this <see cref="DigitalDefinitionBase"/>.
        /// </summary>
        /// <remarks>
        /// Offset for digital values will always be 0; assigning a value other than 0 will thrown an exception.
        /// </remarks>
        /// <exception cref="NotImplementedException">Digital values represent bit flags and thus do not support an offset.</exception>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public sealed override double Offset
        {
            get
            {
                return base.Offset;
            }
            set
            {
                if (value == 0)
                    base.Offset = value;
                else
                    throw new NotImplementedException("Digital values represent bit flags and thus do not support an offset");
            }
        }

        /// <summary>
        /// Gets or sets the integer scaling value of this <see cref="DigitalDefinitionBase"/>.
        /// </summary>
        /// <remarks>
        /// Scaling value for digital values will always be 1; assigning a value other than 1 will thrown an exception.
        /// </remarks>
        /// <exception cref="NotImplementedException">Digital values represent bit flags and thus are not scaled.</exception>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public sealed override uint ScalingValue
        {
            get
            {
                return base.ScalingValue;
            }
            set
            {
                if (value == 1)
                    base.ScalingValue = value;
                else
                    throw new NotImplementedException("Digital values represent bit flags and thus are not scaled");
            }
        }

        /// <summary>
        /// Gets the scale/bit for the <see cref="ScalingValue"/> of this <see cref="DigitalDefinitionBase"/>.
        /// </summary>
        /// <remarks>
        /// Scale/bit for digital values will always be 1.0.
        /// </remarks>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public sealed override double ScalePerBit
        {
            get
            {
                return 1.0D;
            }
        }

        #endregion
    }
}